File: partech/src/rule.rs
#[derive(Hash)]
#[derive(Clone)]
#[derive(Debug)]
#[derive(PartialEq, Eq)]
pub struct Rule<'a> {
    pub name: &'a str,
    pub choices: Vec<(&'a str, Vec<Symbol<'a>>)>,
}

#[derive(Hash)]
#[derive(Clone)]
#[derive(Debug)]
#[derive(PartialEq, Eq)]
pub enum Symbol<'a> {
    Word(&'a str),
    Rule(&'a Rule<'a>),
    Fn(fn () -> Rule<'a>),
    Ap(fn (&'a Vec<Symbol<'a>>) -> Rule, &'a Vec<Symbol<'a>>),
    WordPred(&'a WordPred<'a>),
}

impl <'a> Symbol<'a> {
    pub fn terminal_p(&self) -> bool {
        match self {
            Symbol::Word(_) => { true }
            Symbol::WordPred(_) => { true }
            _ => { false }
        }
    }

    pub fn non_terminal_p(&self) -> bool {
        !self.terminal_p()
    }

    pub fn terminal_match(&self, word: &'a str) -> bool {
        match self {
            Symbol::Word(word2) => { &word == word2 }
            Symbol::WordPred(word_pred) => { (word_pred.pred)(word) }
            _ => {
                panic!("terminal_match should not be called on non terminal");
            }
        }
    }

    pub fn non_terminal_to_rule(&self) -> Rule<'a> {
        match self {
            Symbol::Rule(rule) => { (*rule).clone() }
            Symbol::Fn(f) => { f() }
            Symbol::Ap(f, args) => { f(args) }
            _ => {
                panic!("non_terminal_to_rule should not be called on terminal");
            }
        }
    }
}

#[derive(Hash)]
#[derive(Clone)]
#[derive(Debug)]
#[derive(PartialEq, Eq)]
pub struct WordPred <'a> {
    pub name: &'a str,
    pub pred: fn (&'a str) -> bool,
}



#[test]
fn test_rule_equality() {
    use crate::predefined::non_empty_list;

    fn exp1<'a>() -> Rule<'a> {
        Rule {
            name: "exp1",
            choices: vec![
                ("fn", vec![Symbol::Word("fn")]),
                ("ap", vec![Symbol::Fn(exp1)]),
            ]
        }
    }

    fn exp2<'a>() -> Rule<'a> {
        Rule {
            name: "exp2",
            choices: vec![
                ("fn", vec![Symbol::Word("fn")]),
                ("ap", vec![Symbol::Fn(exp2)]),
            ]
        }
    }

    assert_eq!(exp1(), exp1());
    assert_ne!(exp1(), exp2());

    assert_eq!(
        Symbol::Ap(non_empty_list, &vec![Symbol::Fn(exp1)]),
        Symbol::Ap(non_empty_list, &vec![Symbol::Fn(exp1)]));
    assert_ne!(
        Symbol::Ap(non_empty_list, &vec![Symbol::Fn(exp1)]),
        Symbol::Ap(non_empty_list, &vec![Symbol::Fn(exp2)]));
}

File: src/main/scala/xieyuheng/party/Rule.scala
package xieyuheng.party

case class Rule(name: String, choices: Map[String, List[Symbol]])

sealed trait Symbol {

  def non_terminal_p(): Boolean = {
    this match {
      case SymbolWord(word: String) =>
        false
      case SymbolWordPred(word_pred: WordPred) =>
        false
      case SymbolRule(rule: Rule) =>
        true
      case SymbolFn(fn) =>
        true
      case SymbolAp(fn, args: List[Symbol]) =>
        true
    }
  }

  def non_terminal_to_rule(): Rule = {
    this match {
      case SymbolWord(word: String) =>
        throw new Error(s"non_terminal_to_rule can not handle: ${this}")
      case SymbolWordPred(word_pred: WordPred) =>
        throw new Error(s"non_terminal_to_rule can not handle: ${this}")
      case SymbolRule(rule: Rule) =>
        rule
      case SymbolFn(fn) =>
        fn()
      case SymbolAp(fn, args: List[Symbol]) =>
        fn(args)
    }
  }

  def terminal_p(): Boolean = {
    !this.non_terminal_p()
  }

  def terminal_match(word: String): Boolean = {
    this match {
      case SymbolWord(word2: String) =>
        word == word2
      case SymbolWordPred(word_pred: WordPred) =>
        word_pred.pred(word)
      case SymbolRule(rule: Rule) =>
        throw new Error(s"terminal_match can not handle: ${this}")
      case SymbolFn(fn) =>
        throw new Error(s"terminal_match can not handle: ${this}")
      case SymbolAp(fn, args: List[Symbol]) =>
        throw new Error(s"terminal_match can not handle: ${this}")
    }
  }

  def repr(): String = {
    this match {
      case SymbolWord(word: String) =>
        val doublequote = '"'
        s"${doublequote}${word}${doublequote}"
      case SymbolWordPred(word_pred: WordPred) =>
        s"?${word_pred.name}"
      case SymbolRule(rule: Rule) =>
        s"${rule.name}"
      case SymbolFn(fn) =>
        val rule = fn()
        s"${rule.name}"
      case SymbolAp(fn, args: List[Symbol]) =>
        val rule = fn(args)
        s"${rule.name}"
    }
  }

}

final case class SymbolWord(word: String) extends Symbol
final case class SymbolWordPred(word_pred: WordPred) extends Symbol
final case class SymbolRule(rule: Rule) extends Symbol
final case class SymbolFn(fn: () => Rule) extends Symbol
final case class SymbolAp(fn: List[Symbol] => Rule, args: List[Symbol]) extends Symbol

case class WordPred(name: String, pred: String => Boolean) {

  val matters = name

  override def equals(that: Any): Boolean = {
    that match {
      case that: WordPred => this.matters == that.matters
      case _ => false
    }
  }

  override def hashCode = matters.hashCode

}
